x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichProcessorFactory.java

```diff
@@ -129,15 +131,30 @@ public void accept(ClusterState state) {
     ) {
         Client originClient = new OriginSettingClient(client, ENRICH_ORIGIN);
         return (req, handler) -> {
-            // intentionally non-locking for simplicity...it's OK if we re-put the same key/value in the cache during a race condition.
-            SearchResponse response = enrichCache.get(req);
-            if (response != null) {
-                handler.accept(response, null);
-            } else {
-                originClient.execute(EnrichCoordinatorProxyAction.INSTANCE, req, ActionListener.wrap(resp -> {
-                    enrichCache.put(req, resp);
-                    handler.accept(resp, null);
-                }, e -> { handler.accept(null, e); }));
+            try {
+                CompletableFuture<SearchResponse> cacheEntry = enrichCache.computeIfAbsent(req, request -> {
+                    CompletableFuture<SearchResponse> completableFuture = new CompletableFuture<>();
+                    originClient.execute(
+                        EnrichCoordinatorProxyAction.INSTANCE,
+                        request,
+                        ActionListener.wrap(completableFuture::complete, completableFuture::completeExceptionally)
+                    );
+                    return completableFuture;
+                });
+                cacheEntry.whenComplete((response, throwable) -> {
+                    Exception exception = null;
+                    if (throwable != null) {
+                        enrichCache.invalidate(req, cacheEntry);
+                        if (throwable instanceof Exception) {
+                            exception = (Exception) throwable;
```

Comment:
In case throwable is an instance of `Error` then we should throw it (`throw (Error) throwable;`).
So that the `ElasticsearchUncaughtExceptionHandler` can catch it.

Reply:
Ah yes, a little bit of laziness on my side wasn't sure what was up the call chain and what the implications would be.

Reply:
Hmm this might violate the call back contract - if we throw the error we don't call the callback which might mean we'll hold onto / leak resources. I would make it the callers and the search logic responsible for handling this?

I could of course call the callback and then throw but then we might end up throwing the error twice, if I wrap it and otherwise we lose context. Both don't sound too appealing.

Reply:
`ElasticsearchUncaughtExceptionHandler` will halt the node if a `Error` is encountered. So invoked a callback isn't important, since the jvm will stop. So I think in this case breaking contract is okay.

Wrapping a jvm error into another exception hides potentially bad errors and leave the jvm in a bad state, which is why the in the ES code base `Throwable` is never caught (just `Exception`), so that jvm errors end up in  `ElasticsearchUncaughtExceptionHandler `.

Reply:
Ah excellent. Yeah all bets are off on an error so killing the node is the best we can do. I'll update the code accordingly.

Reply:
Hmm A bit of an edge case, but upon looking at `ElasticsearchUncaughtExceptionHandler` it seems we still deem some throwables as non-fatal meaning we keep the instance around. Opted to report a shallow exception to the callback and then throw the throwable for `ElasticsearchUncaughtExceptionHandler` to handle. If it exits then the callback doesn't matter if the process limps on we try not to make it worse.

Reply:
Addressed in https://github.com/mjmbischoff/elasticsearch/pull/77259/commits/5b9d9bc73cb116248560a036b0fd6559fef85d96